ProductInfoHeader UI 구현

EnnzoKim opened the pull request:
## Screenshots 📸

|구현 화면|
|:-:|
|![image](https://github.com/PyeonHaeng/PyeonHaeng-iOS/assets/97531269/d6b79940-8ba2-4c46-b717-f6938365ef77)|

<br/><br/>

## 고민, 과정, 근거 💬

거의 대부분의 것들이 생소한 것들이라.. 첫 PR은 대략적으로 UI구성만 했습니다. 
`Extensions`같은 필요한 폴더는 일단 임의로 추가하였습니다.

- 1+1, 2+1, 해당없음 이 적혀있는 UI는 아래 컴포넌트를 사용하시면 됩니다.
```swift
PromotionTagView(promotionTag: .onePlus)
```

### UI Tests에 관하여..

UITests 코드를 작성하다가 머리가 터져버렸습니다..
~~테스트 코드 작성하다가 다 지움~~

사실 이번 화면은 사용자의 인터렉션이 없어서 굳이 UI Test를 작성할 필요는 없었다고 생각이 들지만,
**Localizable이 제대로 사용자에게 보여지는가?** 에 대해서 테스트 코드를 작성하는 도중
*'너무 지엽적인 것까지 UITest를 작성하여 시간을 뺏기는 것이 아닌가?'* 라는 생각이 들더라구요..
`예를 들어서 화면 전환도 A화면에서 B화면으로 잘 전환되는가?` 처럼..

혹시 이 부분에 대해서 의견이나 조언같은 것을 주실 수 있다면 감사하겠습니다. _ _

## References 📋

- https://www.pixeldock.com/blog/how-to-use-your-localizable-strings-in-your-xcode-uitests/

<br/><br/>

---

- Closed: #6

EnnzoKim committed:
💄 ProductInfoHeader UI 파일 추가

해당 파일은 제품 상세 페이지 최상단에 있는 HeaderView입니다.

EnnzoKim committed:
✨ 제품의 이미지인 AsyncImage UI 구현

EnnzoKim committed:
💄 ProductInfoHeader, 제품 이름, 가격, 할인 행사 종류 관련 UI 구현제외되어야할 항목

현재 제외되는 항목
– 제품의 카테고리 레이블
– 제품의 원래 가격 레이블

EnnzoKim committed:
✨ 정수형을 천의 단위 원의 기호로 바꿔주는 메서드 구현

EnnzoKim committed:
💄 1+1, 2+1, 해당없음의 관련된 UI, PromotionTagView 구현

EnnzoKim committed:
🌐 "제품상세", "개당", "할인 행사 편의점" 번역

EnnzoKim committed:
💄 컴포넌트 ProductInfoHeader에 삽입 및 적용

EnnzoKim committed:
🎨 자잘한 코드 변경

WhiteHyun commented on PyeonHaeng-iOS/Sources/Scenes/ProductInfoScene/ProductInfoView.swift:
```diff
@@ -7,9 +7,19 @@
 struct ProductInfoView: View {
   var body: some View {
-    Text("")
+    NavigationView {
```
`NavigationView`는 deprecated되어서 `NavigationStack`을 사용해주세요.

WhiteHyun commented on PyeonHaeng-iOS/DesignSystem/Components/PromotionTagView.swift:
```diff
@@ -0,0 +1,47 @@
+
+struct PromotionTagView: View {
+  
+  enum PromotionTag {
+    case onePlus, twoPlus, none
+
+    var string: String {
+      switch self {
+      case .onePlus: "1+1"
+      case .twoPlus: "2+1"
+      case .none: "행사없음"
+      }
+    }
```
PromotionTag라는 열거형이 `PromotionTagView`에만 들어간 이유는 무엇일까요?
저희 디자인을 보았을 때 PromotionTagView 뿐 아니라 다른 곳에서도 사용될 것 같아 보입니다. 따로 모델로 빼는 건 어떻게 생각하시나요?

EnnzoKim replied:
좋습니다! 👍

WhiteHyun commented on PyeonHaeng-iOS/DesignSystem/Components/PromotionTagView.swift:
```diff
@@ -0,0 +1,47 @@
+      case .twoPlus: .pyeonHaengBlue
+      case .none: .gray400
+      }
+    }
```
만약 모델을 분리하게 된다면 Color속성값도 모델이 갖는 게 아니라 TagView가 갖도록 분리할 필요가 있어보입니다. 모델이 뷰의 값을 가지는 것은 좀 어색해보여요.

isGeekCode commented on PyeonHaeng-iOS/Sources/Scenes/ProductInfoScene/ProductInfoHeader.swift:
```diff
@@ -0,0 +1,62 @@
+
+struct ProductInfoHeader: View {
+  /// 임시 데이터입니다.
+  enum MockProduct {
```
enum으로 mock data를 만든건 처음봤어요!!  나중에 상품 구조체가 생기겠죠??

isGeekCode replied:
수고 많으셨어요!

WhiteHyun commented on PyeonHaeng-iOS/Sources/Extensions/Int+.swift:
```diff
@@ -0,0 +1,21 @@
+
+import Foundation
+
+extension Int {
+  /// 천 단위의 Int값에 콤마(,)를 추가합니다.
+  func toStringWithComma() -> String {
+    let numberFormmater = NumberFormatter()
+    numberFormmater.numberStyle = .decimal
+
+    if let numberToString = numberFormmater.string(from: NSNumber(value: self)) {
+      return "\(numberToString)₩"
+    }
+    return ""
+  }
+}
```
약간 수정이 필요한 코드같아 보여요.

### 1. 함수명과 실제 구현체가 다름

thousand 단위마다 `,`를 붙이고 `₩` 기호까지 붙였는데, 메서드 이름과는 좀 다른 것 같습니다. 공통으로 간편하게 사용하기 위한 것이라면 `₩` 기호를 붙이는 건 View에 맡기고, `,`를 붙이는 것, 즉 하나의 역할을 담도록 하는 것은 어떨까요?

### 2. 포매팅에 실패하면 빈 값을 리턴한다?

Int의 `raw data`를 갖고있음에도 빈 값을 리턴하는 이유가 무엇인지 궁금합니다. 저라면 기존 값을 리턴하거나 메서드의 타입을 `String?`으로 변경하고 실패로부터 오는 `부작용(Side-Effect)`는 호출하는 주체에게 맡기는 게 좋다고 생각합니다.

EnnzoKim replied:
의견 감사합니다!

말씀하신대로 `Int`의 raw data로 `return`하는 것은 어떤가요?
`String?`도 하나의 방법이지만, `Text()`는 무조건 `String`만 들어가야해서
SwiftUI의 구조상 언래핑 하는 과정이 조금(?) 더러워지더라고요..!

아래는 수정된 코드입니다!

```swift
extension Int {
  func toStringWithComma() -> String {
    let numberFormmater = NumberFormatter()
    numberFormmater.numberStyle = .decimal
    if let numberToString = numberFormmater.string(from: NSNumber(value: self)) {
      return "\(numberToString)"
    } else {
      os_log(.error, "🚨 %d의 쉼표 표기법 포맷팅을 실패했습니다.", self)
      return "\(self)"
    }
  }
}
```

WhiteHyun replied:
넵 좋습니다. ㅎㅎ
조금 더 첨언하자면 두 가지가 있을 것 같아요 :)

1. deprecated 신경써주세요! 지금 사용하신 os_log 코드는 꽤 예전에 `deprecated`되었습니다. 이건 제가 Shared 내부에 Logger 모듈을 따로 구현해두겠습니다.
2. numberToString을 String interpolation으로 묶지 않고 그냥 리턴해도 될 것 같아요. String타입이니까요!

WhiteHyun reviewed (commented):
사전 PR로 인해 오류가 발생한 것 같아요. conflict 수정 및 제가 요청한 걸 해결하신 뒤 다시 리뷰 요청해 주세요. 그러면 다시 확인하도록 하겠습니다. 고생하셨습니다 ☺️

아직 View가 완벽하게 완성되지는 않았다보니 너무 UITest에 신경쓰지 말아주세요! 제가 말씀드리고 싶었던 것은 커스텀으로 만든 모듈이 있다면 Unit Test로 검증하는 연습을 해보자는 것이었어요. ㅎㅎ

WhiteHyun committed:
:bento: Update colors in Assets.xcassets

- Renamed an existing color to `main` and updated its hue to green.
- Added a new color named `blur` for additional UI elements.

WhiteHyun committed:
:technologist: Extend `Color` for external module access

- Created an extension of `Color` to expose ColorResource to external modules.

WhiteHyun committed:
:technologist: Extend `Image` for external module access

- Created an extension of `Image` to expose ImageResource to external modules

EnnzoKim committed:
🩹 NavigationView -> NavigationStack으로 변경

EnnzoKim committed:
🩹 PromotionTag 전역적 사용을 위해 따로 모델을 분리

EnnzoKim committed:
🩹 PromotionTag 열거형에 필요없는 color프로퍼티 제거 및 View에 해당 코드 추가

EnnzoKim committed:
🩹 toStringWithComma() Return값 변경

EnnzoKim committed:
🚚PromotionTagView, PyeongHaeng-iOS 하위의 Components 폴더로 이동

EnnzoKim committed:
🔀 Merge remote-tracking branch 'origin/feature/designsystem-resource/11' into feature/ProductInfo/6

# Conflicts:
#	PyeonHaeng-iOS.xcodeproj/project.pbxproj

EnnzoKim commented:
@PyeonHaeng/ios 

요청 사항 수정 및 Conflict 수정하여 다시 요청보냅니다!

WhiteHyun commented on PyeonHaeng-iOS/Sources/Components/PromotionTagView.swift:
```diff
@@ -0,0 +1,34 @@
+struct PromotionTagView: View {
+  var promotionTag: PromotionTag
+
+  var color: Color {
+    switch promotionTag {
+    case .onePlus: .pyeonHaengRed
+    case .twoPlus: .pyeonHaengBlue
+    case .none: .gray400
+    }
+  }
```
외부에서 이 값을 접근할 이유가 있을까요?
private로 감싸거나, private extension으로 빼두는 것도 좋을 것 같아요.
후자가 좀 더 가독성에 좋을 것 같네요.
어떻게 생각하시나요?

EnnzoKim replied:
좋습니다! 수정 완료되었습니다~!

WhiteHyun commented on PyeonHaeng-iOS/Sources/PromotionTag.swift:
```diff
@@ -0,0 +1,22 @@
+  case twoPlus
+  case none
+
+  var string: String {
+    switch self {
+    case .onePlus: "1+1"
+    case .twoPlus: "2+1"
+    case .none: "행사없음"
+    }
+  }
```
단순한 string 형식으료 표현하는 거라면 `CustomStringConvertible`라는 것도 있습니다. :)

WhiteHyun commented on PyeonHaeng-iOS/Sources/Scenes/ProductInfoScene/ProductInfoHeader.swift:
```diff
@@ -0,0 +1,63 @@
+          PromotionTagView(promotionTag: .onePlus)
+          Text("개당")
+            .font(.c1)
+          Text("\(MockProduct.price.toStringWithComma())₩")
```
이렇게 하면 Localizable이 자동으로 `%@₩`로 업데이트됩니다. `Text(verbatim:)`를 사용해주세요.

WhiteHyun commented on Shared/Sources/DesignSystem/Sources/Color+Resource.swift:
워우, squash로 넣었다보니 커밋 추적을 못하는군요. 🤣 제 것도 들어가있네요. ㅋㅋ

EnnzoKim committed:
🔥 os_log 삭제 및 String interpolation 삭제

EnnzoKim committed:
🩹color 프로퍼티 private extension로 수정

EnnzoKim committed:
🩹 Text(verbatim:) 으로 수정

EnnzoKim committed:
🩹 CustomStringConvertible 프로토콜 채택

WhiteHyun commented on PyeonHaeng-iOS/Sources/Components/PromotionTagView.swift:
```diff
@@ -0,0 +1,38 @@
+// MARK: - PromotionTagView
+
+struct PromotionTagView: View {
+  var promotionTag: PromotionTag
```
View에서 Tag를 계속 변경하지는 않을 것 같으므로 `private let`으로 접근제어를 제한해도 괜찮다고 생각합니다.
혹여나 다른 플로우가 생길 염려가 있다면 그 때 수정해도 좋을 것 같아요 :)
최대한 빡세게 접근제어를 잡는 쪽으로 고려해봅시다.

EnnzoKim replied:
네 알겠습니다. 일단 이것은 나중에 수정하도록 할게요!

WhiteHyun commented on PyeonHaeng-iOS/Sources/Extensions/Int+.swift:
```diff
@@ -0,0 +1,21 @@
+//
+
+import Foundation
+import OSLog
```
쓰지않는 프레임워크는 제거해주세요.

EnnzoKim committed:
🔥 사용하지 않는 프레임워크 삭제

WhiteHyun reviewed (dismissed):
고생하셨습니다.

isGeekCode reviewed (dismissed):
고생많으셨어요!!

EnnzoKim committed:
💚 Git Action 재시작
